futures-async-macro/src/lib.rs

```diff
@@ -247,6 +222,93 @@ pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
     output.into()
 }
 
+#[proc_macro_attribute]
+pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
+    // Handle arguments to the #[async] attribute, if any
+    let attribute = attribute.to_string();
+    let boxed = if attribute == "( boxed )" {
+        true
+    } else if attribute == "" {
+        false
+    } else {
+        panic!("the #[async] attribute currently only takes `boxed` as an arg");
+    };
+
+    async_inner(boxed, function, quote! { ::futures::__rt::gen }, |output| {
+        // TODO: can we lift the restriction that `futures` must be at the root of
+        //       the crate?
+        let output_span = first_last(&output);
+        let return_ty = if boxed {
+            quote! {
+                Box<::futures::Future<
+                    Item = <! as ::futures::__rt::IsResult>::Ok,
+                    Error = <! as ::futures::__rt::IsResult>::Err,
+                >>
+            }
+        } else {
+            // Dunno why this is buggy, hits weird typecheck errors in tests
+            //
+            // quote! {
+            //     impl ::futures::Future<
+            //         Item = <#output as ::futures::__rt::MyTry>::MyOk,
+            //         Error = <#output as ::futures::__rt::MyTry>::MyError,
+            //     >
+            // }
+            quote! { impl ::futures::__rt::MyFuture<!> + 'static }
+        };
+        let return_ty = respan(return_ty.into(), &output_span);
+        replace_bang(return_ty, &output)
+    })
+}
+
+#[proc_macro_attribute]
+pub fn async_stream(attribute: TokenStream, function: TokenStream) -> TokenStream {
+    // Handle arguments to the #[async_stream] attribute, if any
+    let args = syn::parse::<AsyncStreamArgs>(attribute)
+        .expect("failed to parse attribute arguments");
+
+    let mut boxed = false;
+    let mut item_ty = None;
+
+    for arg in args.0 {
+        match arg {
+            AsyncStreamArg(term, None) => {
+                if term == "boxed" {
```

Comment:
Similar to below, could this panic if `boxed` is already true?

Reply:
done